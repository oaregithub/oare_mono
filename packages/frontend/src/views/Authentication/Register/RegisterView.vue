<template>
  <OareCard :title="$t('register.register')">
    <v-text-field
      outline
      v-model="user.firstname"
      :label="$t('register.firstName')"
      :error-messages="formErrors.firstname"
      outlined
      class="test-firstname"
    ></v-text-field>
    <v-text-field
      outlined
      v-model="user.lastname"
      :label="$t('register.lastName')"
      :error-messages="formErrors.lastname"
      class="test-lastname"
    ></v-text-field>
    <v-text-field
      outlined
      v-model="user.email"
      :label="$t('register.email')"
      :error-messages="formErrors.email"
      class="test-email"
    ></v-text-field>
    <v-text-field
      outlined
      v-model="user.password"
      :label="$t('register.password')"
      :error-messages="formErrors.password"
      type="password"
      class="test-password"
    ></v-text-field>
    <v-text-field
      @keyup.enter="register"
      outlined
      v-model="user.repeatpassword"
      :label="$t('register.confirmPassword')"
      :error-messages="formErrors.repeatpassword"
      name="repeatpassword"
      type="password"
      class="test-confirm-password"
    ></v-text-field>
    <p class="subtitle error--text test-error-msg">{{ errorMsg }}</p>
    <v-btn text class="text-none" to="/login">
      {{ $t('register.alreadyHaveAccount') }}
    </v-btn>

    <template #actions>
      <OareLoaderButton
        color="primary"
        class="test-register-btn"
        @click="register"
        :loading="loading.registerButton"
        :disabled="Object.values(user).some(v => !v.trim())"
      >
        {{ $t('register.confirm') }}
      </OareLoaderButton>
    </template>
  </OareCard>
</template>

<script lang="ts">
import { RegisterPayload } from '@oare/types';
import { defineComponent, ref, PropType } from '@vue/composition-api';
import Router from 'vue-router';
import sl from '@/serviceLocator';
import defaultRouter from '../../../router';

export default defineComponent({
  name: 'RegisterView',
  props: {
    router: {
      type: Object as PropType<Router>,
      default: () => defaultRouter,
    },
  },
  setup({ router }) {
    const server = sl.get('serverProxy');
    const store = sl.get('store');

    const user = ref({
      email: '',
      password: '',
      repeatpassword: '',
      firstname: '',
      lastname: '',
    });

    const formErrors = ref({
      email: '',
      password: '',
      repeatpassword: '',
      firstname: '',
      lastname: '',
    });

    const errorMsg = ref('');
    const loading = ref({
      registerButton: false,
    });

    const validate = () => {
      let valid = true;
      formErrors.value = {
        email: '',
        password: '',
        repeatpassword: '',
        firstname: '',
        lastname: '',
      };

      if (user.value.repeatpassword !== user.value.password) {
        formErrors.value.repeatpassword = 'Passwords do not match';
        valid = false;
      }

      if (user.value.password.length < 8) {
        formErrors.value.password =
          'Password must be at least 8 characters long';
        valid = false;
      }

      if (!/.+@.+\..+/.test(user.value.email)) {
        formErrors.value.email = 'Email is not formatted correctly';
        valid = false;
      }

      return valid;
    };

    const register = async () => {
      loading.value.registerButton = true;

      if (!validate()) {
        loading.value.registerButton = false;
        return;
      }

      errorMsg.value = '';
      let userData: RegisterPayload = {
        firstName: user.value.firstname,
        lastName: user.value.lastname,
        password: user.value.password,
        email: user.value.email,
      };

      try {
        const user = await server.register(userData);
        store.setUser(user);
        router.push('/');
      } catch (e) {
        errorMsg.value = e.response.data.message;
      } finally {
        loading.value.registerButton = false;
      }
    };

    return {
      user,
      errorMsg,
      loading,
      register,
      formErrors,
    };
  },
});
</script>
