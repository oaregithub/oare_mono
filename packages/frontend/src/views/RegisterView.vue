<template>
  <OareUserCard :title="$t('register.register')">
    <template v-slot>
      <v-text-field
        outline
        v-model="user.firstname"
        :label="$t('register.firstName')"
        :error-messages="formErrors.firstname"
        outlined
        class="test-firstname"
      ></v-text-field>
      <v-text-field
        outlined
        v-model="user.lastname"
        :label="$t('register.lastName')"
        :error-messages="formErrors.lastname"
        class="test-lastname"
      ></v-text-field>
      <v-text-field
        outlined
        v-model="user.email"
        :label="$t('register.email')"
        :error-messages="formErrors.email"
        class="test-email"
      ></v-text-field>
      <v-text-field
        outlined
        v-model="user.password"
        :label="$t('register.password')"
        :error-messages="formErrors.password"
        type="password"
        class="test-password"
      ></v-text-field>
      <v-text-field
        @keyup.enter="register"
        outlined
        v-model="user.repeatpassword"
        :label="$t('register.confirmPassword')"
        :error-messages="formErrors.repeatpassword"
        name="repeatpassword"
        type="password"
        class="test-confirm-password"
      ></v-text-field>
      <p class="subtitle error--text test-error-msg">{{ errorMsg }}</p>
      <v-btn text class="text-none" to="/login">
        {{ $t("register.alreadyHaveAccount") }}
      </v-btn>
    </template>

    <template v-slot:actions>
      <OareLoaderButton
        color="primary"
        class="test-register-btn"
        @click="register"
        :loading="loading.registerButton"
        :disabled="Object.values(user).some((v) => !v.trim())"
      >
        {{ $t("register.confirm") }}
      </OareLoaderButton>
    </template>
  </OareUserCard>
</template>

<script lang="ts">
import { defineComponent, ref, Ref, PropType } from '@vue/composition-api';
import Router from 'vue-router';
import { Store } from 'vuex';
import i18n from '../i18n';
import defaultStore from '../store';
import defaultRouter from '../router';

export interface FormRef {
  validate: () => void;
}

export default defineComponent({
  name: 'RegisterView',
  props: {
    router: {
      type: Object as PropType<Router>,
      default: () => defaultRouter,
    },
    store: {
      type: Object as PropType<Store<{}>>,
      default: () => defaultStore,
    },
  },
  setup({ store, router }) {
    const user = ref({
      email: '',
      password: '',
      repeatpassword: '',
      firstname: '',
      lastname: '',
    });

    const formErrors = ref({
      email: '',
      password: '',
      repeatpassword: '',
      firstname: '',
      lastname: '',
    });

    const errorMsg = ref('');
    const loading = ref({
      registerButton: false,
    });

    const validate = () => {
      let valid = true;
      formErrors.value = {
        email: '',
        password: '',
        repeatpassword: '',
        firstname: '',
        lastname: '',
      };

      if (user.value.repeatpassword !== user.value.password) {
        formErrors.value.repeatpassword = 'Passwords do not match';
        valid = false;
      }

      if (user.value.password.length < 8) {
        formErrors.value.password =
          'Password must be at least 8 characters long';
        valid = false;
      }

      if (!/.+@.+\..+/.test(user.value.email)) {
        formErrors.value.email = 'Email is not formatted correctly';
        valid = false;
      }

      return valid;
    };

    const register = async () => {
      loading.value.registerButton = true;

      if (!validate()) {
        loading.value.registerButton = false;
        return;
      }

      errorMsg.value = '';
      let userData = {
        first_name: user.value.firstname,
        last_name: user.value.lastname,
        password: user.value.password,
        email: user.value.email,
      };

      try {
        await store.dispatch('register', userData);
        router.push('/');
      } catch (e) {
        errorMsg.value = e;
      } finally {
        loading.value.registerButton = false;
      }
    };

    return {
      user,
      errorMsg,
      loading,
      register,
      formErrors,
    };
  },
});
</script>
