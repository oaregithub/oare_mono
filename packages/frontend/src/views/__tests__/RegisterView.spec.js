import Vuetify from 'vuetify';
import VueCompositionApi from '@vue/composition-api';
import { mount, createLocalVue } from '@vue/test-utils';
import RegisterView from '../RegisterView.vue';
import flushPromises from 'flush-promises';

const vuetify = new Vuetify();
const localVue = createLocalVue();
localVue.use(VueCompositionApi);

describe('RegisterView test', () => {
  const mockProps = {
    router: {
      push: jest.fn(),
    },
    store: {
      dispatch: jest.fn().mockResolvedValue(true),
    },
  };
  const createWrapper = (props = mockProps) =>
    mount(RegisterView, {
      localVue,
      vuetify,
      propsData: props,
      mocks: {
        $t: jest.fn().mockReturnValue('string'),
      },
      stubs: ['router-link'],
    });

  const fillInForm = async wrapper => {
    const firstNameInput = wrapper.find('.test-firstname input');
    await firstNameInput.setValue('First');

    const lastNameInput = wrapper.find('.test-lastname input');
    await lastNameInput.setValue('Last');

    const emailInput = wrapper.find('.test-email input');
    await emailInput.setValue('email@test.com');

    const passwordInput = wrapper.find('.test-password input');
    await passwordInput.setValue('password');

    const confPasswordInput = wrapper.find('.test-confirm-password input');
    await confPasswordInput.setValue('password');
  };

  const submitForm = async wrapper => {
    const registerBtn = wrapper.find('.test-register-btn');
    await registerBtn.trigger('click');
    await flushPromises();
  };

  it('matches snapshot', () => {
    expect(createWrapper()).toMatchSnapshot();
  });

  it('confirm button is disabled unless all fields are filled in', async () => {
    const wrapper = createWrapper();
    const registerBtn = wrapper.find('.test-register-btn');
    expect(registerBtn.element).toBeDisabled();

    const firstNameInput = wrapper.find('.test-firstname input');
    await firstNameInput.setValue('First');
    expect(registerBtn.element).toBeDisabled();

    const lastNameInput = wrapper.find('.test-lastname input');
    await lastNameInput.setValue('Last');
    expect(registerBtn.element).toBeDisabled();

    const emailInput = wrapper.find('.test-email input');
    await emailInput.setValue('email@test.com');
    expect(registerBtn.element).toBeDisabled();

    const passwordInput = wrapper.find('.test-password input');
    await passwordInput.setValue('password');
    expect(registerBtn.element).toBeDisabled();

    const confPasswordInput = wrapper.find('.test-confirm-password input');
    await confPasswordInput.setValue('password');
    expect(registerBtn.element).not.toBeDisabled();
  });

  it('successfully submits form', async () => {
    const wrapper = createWrapper();
    await fillInForm(wrapper);
    await submitForm(wrapper);

    expect(mockProps.store.dispatch).toHaveBeenCalledWith('register', {
      first_name: 'First',
      last_name: 'Last',
      password: 'password',
      email: 'email@test.com',
    });
    expect(mockProps.router.push).toHaveBeenCalledWith('/');
  });

  it("doesn't submit form if passwords don't match", async () => {
    const wrapper = createWrapper();
    await fillInForm(wrapper);

    const confPasswordInput = wrapper.find('.test-confirm-password input');
    await confPasswordInput.setValue('not-matching');

    await submitForm(wrapper);

    expect(mockProps.store.dispatch).not.toHaveBeenCalled();
    expect(mockProps.router.push).not.toHaveBeenCalled();
  });

  it("doesn't submit if password is less than 8 characters", async () => {
    const wrapper = createWrapper();
    await fillInForm(wrapper);

    const passwordInput = wrapper.find('.test-password input');
    await passwordInput.setValue('short');

    await submitForm(wrapper);

    expect(mockProps.store.dispatch).not.toHaveBeenCalled();
    expect(mockProps.router.push).not.toHaveBeenCalled();
  });

  it("doesn't submit if email is formatted incorrectly", async () => {
    const wrapper = createWrapper();
    await fillInForm(wrapper);

    const emailInput = wrapper.find('.test-email input');
    await emailInput.setValue('bademail');

    await submitForm(wrapper);

    expect(mockProps.store.dispatch).not.toHaveBeenCalled();
    expect(mockProps.router.push).not.toHaveBeenCalled();
  });

  it('shows error message if network calls fail', async () => {
    const mockDispatch = jest.fn().mockRejectedValue('Email already taken');
    const wrapper = createWrapper({
      ...mockProps,
      store: {
        dispatch: mockDispatch,
      },
    });

    await fillInForm(wrapper);
    await submitForm(wrapper);

    const errorMsg = wrapper.find('.test-error-msg');
    expect(errorMsg.text()).toBe('Email already taken');
    expect(mockProps.store.dispatch).not.toHaveBeenCalled();
    expect(mockProps.router.push).not.toHaveBeenCalled();
  });
});
