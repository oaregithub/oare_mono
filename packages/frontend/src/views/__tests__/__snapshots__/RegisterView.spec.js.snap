// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterView test matches snapshot 1`] = `
<div class="mt-12 pa-3 mx-auto v-card v-sheet theme--light elevation-24" style="max-width: 600px;">
  <div class="v-card__title">
    <h3 class="oare-header mb-2">string</h3>
  </div>
  <div class="v-card__text">
    <div class="v-input test-firstname theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
      <div class="v-input__control">
        <div class="v-input__slot">
          <fieldset aria-hidden="true">
            <legend style="width: 0px;"><span>​</span></legend>
          </fieldset>
          <div class="v-text-field__slot"><label for="input-6" class="v-label theme--light" style="left: 0px; position: absolute;">string</label><input outline="" id="input-6" type="text"></div>
        </div>
        <div class="v-text-field__details">
          <div class="v-messages theme--light">
            <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
          </div>
        </div>
      </div>
    </div>
    <div class="v-input test-lastname theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
      <div class="v-input__control">
        <div class="v-input__slot">
          <fieldset aria-hidden="true">
            <legend style="width: 0px;"><span>​</span></legend>
          </fieldset>
          <div class="v-text-field__slot"><label for="input-9" class="v-label theme--light" style="left: 0px; position: absolute;">string</label><input id="input-9" type="text"></div>
        </div>
        <div class="v-text-field__details">
          <div class="v-messages theme--light">
            <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
          </div>
        </div>
      </div>
    </div>
    <div class="v-input test-email theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
      <div class="v-input__control">
        <div class="v-input__slot">
          <fieldset aria-hidden="true">
            <legend style="width: 0px;"><span>​</span></legend>
          </fieldset>
          <div class="v-text-field__slot"><label for="input-12" class="v-label theme--light" style="left: 0px; position: absolute;">string</label><input id="input-12" type="text"></div>
        </div>
        <div class="v-text-field__details">
          <div class="v-messages theme--light">
            <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
          </div>
        </div>
      </div>
    </div>
    <div class="v-input test-password theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
      <div class="v-input__control">
        <div class="v-input__slot">
          <fieldset aria-hidden="true">
            <legend style="width: 0px;"><span>​</span></legend>
          </fieldset>
          <div class="v-text-field__slot"><label for="input-15" class="v-label theme--light" style="left: 0px; position: absolute;">string</label><input id="input-15" type="password"></div>
        </div>
        <div class="v-text-field__details">
          <div class="v-messages theme--light">
            <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
          </div>
        </div>
      </div>
    </div>
    <div class="v-input test-confirm-password theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
      <div class="v-input__control">
        <div class="v-input__slot">
          <fieldset aria-hidden="true">
            <legend style="width: 0px;"><span>​</span></legend>
          </fieldset>
          <div class="v-text-field__slot"><label for="input-18" class="v-label theme--light" style="left: 0px; position: absolute;">string</label><input name="repeatpassword" id="input-18" type="password"></div>
        </div>
        <div class="v-text-field__details">
          <div class="v-messages theme--light">
            <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
          </div>
        </div>
      </div>
    </div>
    <p class="subtitle error--text test-error-msg"></p>
    <router-link-stub to="/login" tag="a" activeclass="v-btn--active" exactactiveclass="v-btn--active" ariacurrentvalue="page" event="click" class="text-none v-btn v-btn--flat v-btn--router v-btn--text theme--light v-size--default"><span class="v-btn__content">
      string
    </span></router-link-stub>
  </div>
  <div class="v-card__actions"><button type="button" disabled="disabled" class="test-register-btn v-btn v-btn--contained v-btn--disabled theme--light v-size--default" color="primary"><span class="v-btn__content">
      string
    </span></button></div>
</div>
`;
